### PR TITLE
Automate Reference Documentation for Configuration Files

### PR DESCRIPTION
**Pull request by nipunisalgado**
## Purpose
Currently, in the Wiki documentation documenting of the Configuration Files of WSO2 products are done manually. It leads to issues such as maintenance headaches and missing content. Therefore, purpose of this feature is to automate a process that extracts this content from the source code of the configuration files and build an HTML output that users can easily browse and search.  

## Goals
When migrating from C4 to C5, instead of XML files YAML files are used as the configuration files. There is already a plugin (config-docs plugin) built to extract the content from the configuration bean classes and to automatically generate a YAML file. The goal of this feature is to extend this plugin to generate an HTML file by using the extracted content. 

## Approach
The implementation of the solution is done by using Handlebars.java.

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java [hunk 1]
```diff
@@ -124,15 +145,24 @@ public void execute() throws MojoExecutionException, MojoFailureException {
                             .description()));
                     // add root node to the config Map
                     finalMap.put(configuration.namespace(), readConfigurationElements(configObject, Boolean.TRUE));
+                    contentMap.put(configuration.namespace(), readConfigurationElements(configObject, Boolean.TRUE));
```
Comment (daneshk) [inline extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java]: Why are we need to call readConfigurationElement method again here. and why we need another map(contentMap). can't we use the same map.

File: extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java [hunk 2]
```diff
@@ -124,15 +145,24 @@ public void execute() throws MojoExecutionException, MojoFailureException {
...
                     // write configuration map as a yaml file
                     writeConfigurationFile(finalMap, configuration.namespace());
+                    readCSS();
+                    copyCSS(readCSS());
+                    //List<DataModel> configItems = new ArrayList<DataModel>();
+                    contextMap.put(getYamlString(contentMap), generateContent(contentMap, configItems));
```
Comment (daneshk) [inline extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java]: As discussed, remove the context map and create new bean class for that.

File: extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java [hunk 3]
```diff
@@ -108,7 +123,13 @@ public void execute() throws MojoExecutionException, MojoFailureException {
 
         // process configuration bean to create configuration document
         for (String configClassName : configurationClasses) {
+            // configuration map used to generate the YAML file
             Map<String, Object> finalMap = new LinkedHashMap<>();
+            // configuration map used to create the object model
+            Map<String, Object> contentMap = new LinkedHashMap<>();
+            // map object passed to the handlebars template
+            //Map<String, List<ConfigItem>> contextMap = new LinkedHashMap<>();
+            Map<String, List<DataModel>> contextMap = new LinkedHashMap<>();
```
Comment (daneshk) [inline extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java]: We don't need to both maps. can we remove them

File: extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java [hunk 4]
```diff
@@ -124,15 +145,24 @@ public void execute() throws MojoExecutionException, MojoFailureException {
                             .description()));
                     // add root node to the config Map
                     finalMap.put(configuration.namespace(), readConfigurationElements(configObject, Boolean.TRUE));
+                    contentMap.put(configuration.namespace(), readConfigurationElements(configObject, Boolean.TRUE));
                     // write configuration map as a yaml file
                     writeConfigurationFile(finalMap, configuration.namespace());
+                    readCSS();
+                    copyCSS(readCSS());
```
Comment (daneshk) [inline extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java]: why do you call readCSS() again. can we remove one.

File: extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java [hunk 5]
```diff
@@ -189,14 +220,379 @@ private void writeConfigurationFile(Map<String, Object> finalMap, String filenam
...
+        StringBuilder sb = new StringBuilder();
+
+        try (BufferedReader br = new BufferedReader(new InputStreamReader(inputStream, UTF_8_CHARSET))) {
+            String line;
+            while ((line = br.readLine()) != null) {
+                sb.append(line).append("\n");
+            }
+        } catch (IOException e) {
+            logger.error("Error while reading the css file.", e);
+        }
```
Comment (daneshk) [inline extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java]: InputStream not handled property. we need to close the stream in the finally block

File: extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java [hunk 6]
```diff
@@ -189,14 +220,379 @@ private void writeConfigurationFile(Map<String, Object> finalMap, String filenam
...
+                }
+
+            }
+
+
+        }
+
+        return configItems;
+
+    }*/
```
Comment (daneshk) [inline extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java]: Why this code is commented out. please remove if we don't use it.

File: extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java [hunk 7]
```diff
@@ -189,14 +220,379 @@ private void writeConfigurationFile(Map<String, Object> finalMap, String filenam
...
+                    generateContent((Map<String, Object>) entry.getValue(), dataModel.childElements);
+                }
+
+            }
+
+
+        }
+
+        return configItems;
+
```
Comment (daneshk) [inline extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java]: Remove blank lines

File: extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ProductConfigurationMojo.java [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+        }
+
+        Context context = Context.newContext(menuItems);
+
+        String html = template.apply(context);
+
+        File configDir = new File(outputDirectory.getParent(), CONFIG_DOCS_DIR);
+
+
+        // create config directory inside project output directory to save config files
```
Comment (daneshk) [inline extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ProductConfigurationMojo.java]: Remove unnecessary new lines.

File: samples/config-generator/src/main/java/org/wso2/carbon/config/samples/ParentConfiguration.java [hunk 9]
```diff
@@ -60,6 +64,10 @@ public ChildConfiguration getChildConfiguration() {
         return childConfiguration;
     }
 
+    /*public MetricsLevelConfig getLevels() {
+        return levels;
+    }*/
+
```
Comment (daneshk) [inline samples/config-generator/src/main/java/org/wso2/carbon/config/samples/ParentConfiguration.java]: Remove commented codes

File: samples/config-generator/src/main/java/org/wso2/carbon/config/samples/ParentConfiguration.java [hunk 10]
```diff
@@ -28,10 +28,13 @@
  *
  * @since 1.0.0
  */
-@Configuration(namespace = "wso2.configuration", description = "Parent configuration")
+@Configuration(namespace = "wso2.configuration", displayName = "Parent Configurations",
+        description = "Parent configuration")
 public class ParentConfiguration {
 
-    @Element(description = "An example element for this configuration")
+    /*private MetricsLevelConfig levels = new MetricsLevelConfig();*/
```
Comment (daneshk) [inline samples/config-generator/src/main/java/org/wso2/carbon/config/samples/ParentConfiguration.java]: Remove commented codes

File: samples/config-generator/src/main/java/org/wso2/carbon/config/samples/ParentConfiguration.java [hunk 11]
```diff
@@ -44,6 +47,7 @@
     @Element(description = "Second level configuration")
     private ChildConfiguration childConfiguration = new ChildConfiguration();
 
+
```
Comment (daneshk) [inline samples/config-generator/src/main/java/org/wso2/carbon/config/samples/ParentConfiguration.java]: Remove unnecessary lines

File: extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ProductConfigurationMojo.java [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+        Resource resource = new Resource();
+        resource.setDirectory(configDir.getAbsolutePath());
+        resource.setTargetPath(CONFIG_DOCS_DIR);
+        project.addResource(resource);
+
+
+    }
+
+
+}
```
Comment (daneshk) [inline extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ProductConfigurationMojo.java]: remove unnecessary lines

File: extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ProductConfigurationMojo.java [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+            generateIndex(generateMenu(menuMap, new File(outputDirectory.getParent(), CONFIG_DOCS_DIR)));
+        } catch (IOException e) {
+            throw new MojoExecutionException("Error while executing the method", e);
+        }
+    }
+
+    private Map<String, String> generateMenu(Map<String, String> menuMap, File directory)
+            throws MojoExecutionException {
+        File configDirt = directory;
+        File dir = new File(configDirt.getPath());
```
Comment (daneshk) [inline extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ProductConfigurationMojo.java]: Can directory file be null? We don't need to assign to new variable

File: extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ProductConfigurationMojo.java [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+        Map<String, String> menuMap = new LinkedHashMap<>();
+
+        try {
+            generateIndex(generateMenu(menuMap, new File(outputDirectory.getParent(), CONFIG_DOCS_DIR)));
+        } catch (IOException e) {
+            throw new MojoExecutionException("Error while executing the method", e);
+        }
+    }
+
+    private Map<String, String> generateMenu(Map<String, String> menuMap, File directory)
```
Comment (daneshk) [inline extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ProductConfigurationMojo.java]: Why we need to send empty map as menuMap as an input parameter. We can remove it.

File: extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ProductConfigurationMojo.java (hunk 13 above)
Comment (daneshk) [reply extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ProductConfigurationMojo.java]: dir is not a good variable name. variable name should be self descriptive.

File: extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java [hunk 15]
```diff
@@ -119,20 +139,30 @@ public void execute() throws MojoExecutionException, MojoFailureException {
...
                     Object configObject = configClass.newInstance();
                     System.clearProperty(ConfigConstants.SYSTEM_PROPERTY_DOC_GENERATION);
+                    Object configElements = readConfigurationElements(configObject, Boolean.TRUE);
                     // add description comment to the root node.
                     finalMap.put(COMMENT_KEY_PREFIX + configuration.namespace(), createDescriptionComment(configuration
                             .description()));
                     // add root node to the config Map
-                    finalMap.put(configuration.namespace(), readConfigurationElements(configObject, Boolean.TRUE));
+                    finalMap.put(configuration.namespace(), configElements);
+                    contentMap.put(configuration.namespace(), configElements);
```
Comment (daneshk) [inline extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java]: why we need contentMap. can't we use the same finalMap

**Comment by daneshk** [comment]
Please add unit testcases for new classes and methods.

File: extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java [hunk 16]
```diff
@@ -119,20 +139,30 @@ public void execute() throws MojoExecutionException, MojoFailureException {
...
                     // write configuration map as a yaml file
                     writeConfigurationFile(finalMap, configuration.namespace());
+                    copyCSS(readCSS());
+                    //copyJS(readJS());
```
Comment (daneshk) [inline extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java]: Remove commented out codes.

File: extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java [hunk 17]
```diff
@@ -188,14 +218,248 @@ private void writeConfigurationFile(Map<String, Object> finalMap, String filenam
...
+    /**
+     * copy css file in the resource folder.
+     *
+     * @throws MojoExecutionException
+     */
+    private void copyCSS(String css) throws MojoExecutionException {
+
+        File configDir = new File(project.getBuild().getOutputDirectory(), ConfigConstants.CONFIG_DIR);
+
+        if (!configDir.exists() && !configDir.mkdirs()) {
```
Comment (daneshk) [inline extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java]: Remove new lines

File: extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java [hunk 18]
```diff
@@ -188,14 +218,248 @@ private void writeConfigurationFile(Map<String, Object> finalMap, String filenam
...
+        String description = null;
+        String dataType = null;
+        String defaultValue = null;
+        String required = null;
+        String possibleValues = null;
+
+        for (Map.Entry<String, Object> entry : contentMap.entrySet()) {
+
+            if (entry.getKey().contains(COMMENT_KEY_PREFIX)) {
+
```
Comment (daneshk) [inline extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java]: remove empty lines

File: extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java [hunk 19]
```diff
@@ -188,14 +218,248 @@ private void writeConfigurationFile(Map<String, Object> finalMap, String filenam
...
+                    dataModel.setDataType(dataType);
+                    dataModel.setDescription(description);
+                    dataModel.setDefaultValue(defaultValue);
+                    dataModel.setRequired(required);
+                    dataModel.setPossibleValues(possibleValues);
+
+                    dataType = "";
+                    defaultValue = "";
+                    possibleValues = "";
+
```
Comment (daneshk) [inline extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java]: We can move this initialize at the begin of the iteration

File: extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java [hunk 20]
```diff
@@ -188,14 +218,248 @@ private void writeConfigurationFile(Map<String, Object> finalMap, String filenam
...
+                    dataModel.setElementName(elementName);
+                    dataModel.setDataType(dataType);
+                    dataModel.setDescription(description);
+                    dataModel.setDefaultValue(defaultValue);
+                    dataModel.setRequired(required);
+                    dataModel.setPossibleValues(possibleValues);
+
+                    dataType = "";
+                    defaultValue = "";
+                    possibleValues = "";
```
Comment (daneshk) [inline extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java]: We can move this initialize at the begin of the iteration

File: extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java [hunk 21]
```diff
@@ -188,14 +218,248 @@ private void writeConfigurationFile(Map<String, Object> finalMap, String filenam
...
+                }
+
+            }
+
+
+        }
+
+        return configItemsList;
+
+    }
```
Comment (daneshk) [inline extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java]: Remove unnessary new lines

File: extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java [hunk 22]
```diff
@@ -188,14 +218,248 @@ private void writeConfigurationFile(Map<String, Object> finalMap, String filenam
...
+
+        File configDir = new File(project.getBuild().getOutputDirectory(), ConfigConstants.CONFIG_DIR);
+
+
+        // create config directory inside project output directory to save config files
+        if (!configDir.exists() && !configDir.mkdirs()) {
+            throw new MojoExecutionException("Error while creating config directory in classpath");
+        }
+
+        try (PrintWriter out = new PrintWriter(new File(configDir.getPath(), filename
```
Comment (daneshk) [inline extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java]: Remove unnessary new lines

File: extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/DataModel.java [hunk 23]
```diff
@@ -0,0 +1,78 @@
+package org.wso2.carbon.config.maven.plugin;
```
Comment (daneshk) [inline extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/DataModel.java]: License header is missing

File: extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/DataModel.java [hunk 24]
```diff
@@ -0,0 +1,78 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * This class will define the structure of the object model
+ * which is passed as the context to the handlebars template
+ * 'main.hbs' to generate the content displayed in the
+ * HTML files.
+ *
+ * @since 1.0.0
```
Comment (daneshk) [inline extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/DataModel.java]: version 1.0.0 is already released. set the next release version

File: extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java (hunk 16 above)
Comment (nipunisalgado) [reply extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java]: Done

File: extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java (hunk 17 above)
Comment (nipunisalgado) [reply extensions/org.wso2.carbon.config.maven.plugin/src/main/java/org/wso2/carbon/config/maven/plugin/ConfigDocumentMojo.java]: Done

**Pull request closed without merging**